### PR TITLE
python312Packages.inotify: drop nose dependency; modernize

### PR DESCRIPTION
**Pull request by pyrox0**
## Description of changes
Part of #326513

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by emilazy — COMMENTED** [review]
Looks like upstream is [pretty](https://github.com/dsoprea/PyInotify/issues) [dead](https://github.com/dsoprea/PyInotify/pulls). The only user of this in‐tree is `lxd-image-server`, which hasn’t been updated since it was first added, but it does have a NixOS module…

@mkg20001 @aanderse Hi! Do you still use `lxd-image-server`? Is it still relevant in the Incus era?

**Comment by aanderse** [comment]
not personally using it, though maybe there are others that are

cc @adamcstephens

**Comment by adamcstephens** [comment]
I have never used `lxd-image-server`, but assuming it works with LXD I would expect it to work with Incus. They share the same remote image protocol.

File: pkgs/development/python-modules/inotify/default.nix [hunk 1]
```diff
@@ -2,31 +2,41 @@
...
   };
 
-  nativeCheckInputs = [ nose ];
+  postPatch = ''
+    # First is needed because assertEquals was removed in python 3.12
+    # Second and third are needed because these are the opposite pair now for some reason
+    substituteInPlace tests/test_inotify.py \
+      --replace-fail "assertEquals" "assertEqual" \
+      --replace-fail "['IN_ISDIR', 'IN_CREATE']" "['IN_CREATE', 'IN_ISDIR']" \
+      --replace-fail "['IN_ISDIR', 'IN_DELETE']" "['IN_DELETE', 'IN_ISDIR']"
```
[the remark this answers is not in the document]
Comment (emilazy) [reply pkgs/development/python-modules/inotify/default.nix:27]: Can we disable these flaky tests, if that’s what’s going on?

[the remark this answers is not in the document]
Comment (emilazy) [reply pkgs/development/python-modules/inotify/default.nix:27]: @pyrox0 Would be happy to merge this if we go with disabling any nondeterministic tests :)

[the remark this answers is not in the document]
Comment (pyrox0) [reply pkgs/development/python-modules/inotify/default.nix:27]: This fell off my radar the last few weeks, lemme disable those tests.

[the remark this answers is not in the document]
Comment (emilazy) [reply pkgs/development/python-modules/inotify/default.nix]: No worries, I’ve been on and off with this stuff myself.

Result of `nixpkgs-review pr 330776` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>lxd-image-server</li>
    <li>lxd-image-server.dist</li>
    <li>pyrosimple</li>
    <li>pyrosimple.dist</li>
    <li>python311Packages.inotify</li>
    <li>python311Packages.inotify.dist</li>
    <li>python312Packages.inotify</li>
    <li>python312Packages.inotify.dist</li>
  </ul>
</details>

LGTM; could you add a comment explaining why the tests are disabled?

[the remark this answers is not in the document]
Comment (pyrox0) [reply pkgs/development/python-modules/inotify/default.nix]: Updated with an appropriate comment.

**Review by emilazy — APPROVED** [review]
Thanks for all your great work on this project; we couldn’t have done it without you and it looks like we’re on track to remove nose for 24.11 :) I’m looking into the remaining two packages without open PRs, but if I don’t get around to it in the next day or two you can feel free to pick them up if you feel like it.

**Pull request merged**